### ISSUE TITLE
Subs needs to be more aware of the expr it is working with

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#15226]
One of the motivations for Subs was to be able to handle Derivatives with expressions replacing functions. Right now, however, it doesn't process its arguments in a way that one might expect:

```
>>> Subs(Derivative(f(y), x, y), y, g(x)).doit()
0  <-- should be Derivative(f(g(x)), g(x))*Derivative(g(x), x)
          or Subs(Derivative(f(y), y), y, g(x))*Derivative(g(x), x)
```

N.B.
If you try to confirm that assertions above with master you may see
```
>>> Derivative(Subs(Derivative(f(y), y), y, g(x)),x).doit()
Derivative(f(g(x)), (g(x), 2))*Derivative(g(x), x)
```
That's another issue.

**Comment by asmeurer** [comment]
Shouldn't Subs(expr, old new).doit() always be equivalent to expr.subs(old, new)?

**Comment by asmeurer** [comment]
I mean `doit` should literally just call `subs` and do nothing else. Is there need for other code in there that could make the results different?

**Comment by smichr** [comment]
So if you have `Subs(Derivative(f(x), x), (x, f), (x**2, Lambda(x, exp(x))))` you can't just do the subs then doit because you can't substitute in the `x**2`...and you can't doit and substitute in the variables because there is nothing to do. You have to substitute what you can and then do when you can and then finish the substitution.

**Comment by tBuLi** [comment]
I think I found a related instance:
```python
>>> dz = Derivative(z(y(x, a, b), a), a).doit()
Derivative(y(x, a, b), a)*Subs(Derivative(z(_xi_1, a), _xi_1), (_xi_1,), (y(x, a, b),)) + Subs(Derivative(z(y(x, a, b), _xi_2), _xi_2), (_xi_2,), (a,))
>>> for subs in dz.atoms(sympy.Subs):
>>>     dz = dz.xreplace({subs: subs.doit()})
>>> dz
Derivative(y(x, a, b), a)*Subs(Derivative(z(_xi_1, a), _xi_1), (_xi_1,), (y(x, a, b),)) + Subs(Derivative(z(y(x, a, b), _xi_2), _xi_2), (_xi_2,), (a,))
```
Nothing has happened. The Interesting things is that if I use `dz = Derivative(z(y(x, a, b)), a).doit()` instead, it does work. (So `z` is no longer directly a function of `a`, although `y` is.)

If I manually replace the `_xi_1` in `z(_xi_1, a), _xi_1)`, and then call `doit` on the `Subs`, I also get an incorrect zero.

**Comment by smichr** [comment]
The OP expression giving 0 was fixed in a521b2d93a141f91b020456fceb1bec15fbed58b but the added test was incorrectly numbered as 15266 instead of 15226. 

The test could be modified to read
```python
assert Subs(Derivative(f(y), x, y), y, g(x)).doit() == Derivative(f(g(x)), (g(x), 2))*Derivative(g(x), x)
```
I checked it (now) like this:
```python
>>> s=Subs(Derivative(f(y), x, y), y, g(x))
>>> y1=s.doit().subs(f, Lambda(x, sqrt(x))).subs(g(x),x**3).doit()
>>> y2=s.subs(f, Lambda(x, sqrt(x))).subs(g(x),x**3).doit()
>>> y1.equals(y2)
True
```
